File: lc/A084_Largest_Rectangle_in_Histogram.js
/**
 * Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.
 */




const largestRectangleArea = (heights) => {
    // Different from LC11. Container with most water
    let maxArea = 0;
    let monoStack = [];
    for (let i = 0; i < heights.length; i++) {
        while (monoStack.length > 0 && heights[i] < heights[monoStack[monoStack.length - 1]]) {
            let idx = monoStack.pop();
            let left = monoStack.length > 0 ? monoStack[monoStack.length - 1] : -1; 
            maxArea = Math.max(maxArea, (i - left - 1) * heights[idx]);
        }
        monoStack.push(i);
    }
    return maxArea;
}

console.log(largestRectangleArea([2,1,5,6,2,3]))

File: java/src/leetcode/problems/A084_Largest_Rectangle_in_Histogram.java
package leetcode.problems;

import java.util.Stack;

public class A084_Largest_Rectangle_in_Histogram {

    public int largestRectangleArea(int[] heights) {
        if (heights == null || heights.length == 0) return 0;
        int res = 0;
        Stack<Integer> stack = new Stack<>();
        int i = 0;
        while (i < heights.length) {
            if (stack.isEmpty() || heights[i] >= heights[stack.peek()]) {
                stack.push(i++);
            } else {
                int idx = stack.pop();
                int left = stack.isEmpty() ? -1 : stack.peek();
                res = Math.max(res, (i - left - 1) * heights[idx]);
            }
        }
        while (!stack.isEmpty()) {
            int idx = stack.pop();
            int left = stack.isEmpty() ? -1 : stack.peek();
            res = Math.max(res, (i - left - 1) * heights[idx]);
        }

        return res;
    }
}
